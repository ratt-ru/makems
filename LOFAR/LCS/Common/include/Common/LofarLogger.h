//#  LofarLogger.h: Macro interface to the lofar logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: LofarLogger.h 11806 2008-09-28 21:07:32Z loose $

#ifndef LOFAR_COMMON_LOFARLOGGER_H
#define LOFAR_COMMON_LOFARLOGGER_H

// \file
// Macro interface to the lofar logging package.

//# Include the correct set of macros, depending on the availability of the
//# log4cplus or log4cxx packages.
#if defined(HAVE_LOG4CPLUS)
# include <Common/LofarLog4Cplus.h>
#elif defined(HAVE_LOG4CXX)
# include <Common/LofarLog4Cxx.h>
#else
# include <Common/LofarLogCout.h>
#endif

#include <Common/Exceptions.h>

namespace LOFAR 
{

// \ingroup Common
// \addtogroup LofarLogger
  // @{


//# --------------------- Common definitions ---------------
//
// \name Common definitions 
// @{

// The package name of the current package. This variable should be defined in
// the file lofar_config.h which is generated by the lofarconf script. Here,
// we will make sure that LOFARLOGGER_PACKAGE is always defined.
#ifndef LOFARLOGGER_PACKAGE
# define LOFARLOGGER_PACKAGE "Unknown_package"
#endif

// The full package name is defined as the concatenation of the package name
// and an optional sub-package name that may be defined by the user.
#ifdef LOFARLOGGER_SUBPACKAGE
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE "." LOFARLOGGER_SUBPACKAGE
#else
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE
#endif

// @}

//#------------------------ Assert and FailWhen -------------------------------
//#
//# NB: THROW is defined in package-dependant file.
//
// \name Assert and FailWhen
// \note The \c DBG version of \c ASSERT and \c FAILWHEN will only be in your
// compiled code when the (pre)compiler flag \c ENABLE_DBGASSERT is defined.
// @{

// If the condition of assert is NOT met a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERT(cond) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, "Assertion: " #cond); \
	} while(0)


// If the condition of assert is NOT met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERTSTR(cond,stream) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, \
			   "Assertion: " #cond "; " << stream); \
	} while(0)

// If the condition of failwhen is met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define FAILWHEN(cond) do { \
	if (cond) THROW(::LOFAR::AssertError, "Failtest: " #cond); \
	} while(0)

#ifdef ENABLE_DBGASSERT
#define DBGASSERT(cond)				ASSERT(cond)
#define DBGASSERTSTR(cond,stream)	ASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)			FAILWHEN(cond)
#else
#define DBGASSERT(cond)
#define DBGASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)
#endif

// @}
    
} // namespace LOFAR

#endif // file read before
