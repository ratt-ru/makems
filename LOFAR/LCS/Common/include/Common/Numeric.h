//#  Numeric.h: compare floating point numbers.
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: Numeric.h 8240 2006-04-28 12:04:29Z loose $

#ifndef LOFAR_COMMON_NUMERIC_H
#define LOFAR_COMMON_NUMERIC_H

//# Never #include <config.h> or #include <lofar_config.h> in a header file!

// \file
// Compare floating point numbers.

// \def USE_INFINITYCHECK
// Add check for infinity. You will need this check if you don't want infinity
// to be treated as "close to" \c FLT_MAX.
#define USE_INFINITYCHECK

// \def USE_NANCHECK
// Add check for Not-a-Number. You will need this check if you want to make
// sure that two NANs are not treated as equal or "close to" one another; or
// if you want to use very large values for maxUlps.
#define USE_NANCHECK

//# \def USE_SIGNCHECK
//# Add check for sign. You will need this check if you don't want a very small
//# positive number to compare as "close to" a very small negative number.
/* #define USE_SIGNCHECK */
#undef USE_SIGNCHECK

namespace LOFAR
{
  // \addtogroup Common
  // @{

  // Numeric operations.
  // \attention These operations are only valid for machines that support the
  // IEEE 754 floating-point standard. Unfortunately, there is no easy way to
  // check this at compile-time.
  class Numeric
  {
  public:
    // Mask type used for floats is a (32-bit) int.
    typedef int floatMask_t;
    // Mask type used for doubles is a (64-bit) long long.
    typedef long long doubleMask_t;

    // \return \c true if f < 0; otherwise \c false.
    static bool isNegative(float f);
    // \return \c true if d < 0; otherwise \c false.
    static bool isNegative(double d);

    // \return \c true if \a f is neither inifinite nor "not-a-number" (NaN);
    // otherwise \c false.
    static bool isFinite(float f);
    // \return \c true if \a d is neither inifinite nor "not-a-number" (NaN);
    // otherwise \c false.
    static bool isFinite(double d);

    // \return \c true if \a f is infinite; otherwise \c false.
    static bool isInf(float f);
    // \return \c true if \a d is infinite; otherwise \c false.
    static bool isInf(double d);

    // \return \c true if \a f is "not-a-number" (NaN); otherwise \c false.
    static bool isNan(float f);
    // \return \c true if \a f is "not-a-number" (NaN); otherwise \c false.
    static bool isNan(double d);

    // \name Compare 
    // Compare two numeric values. \a lhs and \a rhs are considered
    // equal when the distance, measured in ULPs (Units in Last Place), between
    // them is less than \a maxUlps.
    //@{
    static bool compare(float lhs, float rhs, floatMask_t maxUlps = 8);
    static bool compare(double lhs, double rhs, doubleMask_t maxUlps = 32);
    // @}

  private:

    // @{
    // These constants are for IEEE754 floating point numbers.
    static const floatMask_t floatNegativeMask = 0x80000000L;
    static const floatMask_t floatExponentMask = 0x7F800000L;
    static const floatMask_t floatMantissaMask = 0x007FFFFFL;

    static const doubleMask_t doubleNegativeMask = 0x8000000000000000LL;
    static const doubleMask_t doubleExponentMask = 0x7FF0000000000000LL;
    static const doubleMask_t doubleMantissaMask = 0x000FFFFFFFFFFFFFLL;
    // @}

  };

} // namespace LOFAR

#endif
