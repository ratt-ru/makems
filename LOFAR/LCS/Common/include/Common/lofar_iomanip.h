//#  lofar_iomanip.h:
//#
//#  Copyright (C) 2002
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: lofar_iomanip.h 7330 2005-12-27 10:02:33Z loose $

#ifndef LOFAR_COMMON_IOMANIP_H
#define LOFAR_COMMON_IOMANIP_H

// \file

#include <iomanip>

namespace LOFAR
{
  using std::boolalpha;
  using std::noboolalpha;
  using std::showbase;
  using std::noshowbase;
  using std::showpoint;
  using std::noshowpoint;
  using std::showpos;
  using std::noshowpos;
  using std::skipws;
  using std::noskipws;
  using std::uppercase;
  using std::nouppercase;
  using std::internal;
  using std::left;
  using std::right;
  using std::dec;
  using std::hex;
  using std::oct;
  using std::fixed;
  using std::scientific;
  using std::resetiosflags;
  using std::setiosflags;
  using std::setbase;
  using std::setfill;
  using std::setprecision;
  using std::setw;
}

#endif
