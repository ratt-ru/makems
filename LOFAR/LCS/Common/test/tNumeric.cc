//#  tNumeric.cc: test program for the Numeric class
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: tNumeric.cc 8994 2006-07-24 11:58:20Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <Common/Numeric.h>
#include <Common/LofarLogger.h>
#include <Common/lofar_iomanip.h>
#include <limits>
#include <cmath>

using namespace LOFAR;
using namespace std;

#define initNumbers(T)                                               \
  LOG_INFO("initNumbers("#T")");                                     \
  typedef Numeric::T##Mask_t mask_t;                                 \
  ASSERT(sizeof(T) == sizeof(mask_t));				     \
  T zero(0), one(1), two(2);                                         \
  /* Create a negative zero                              */          \
  T negativeZero;                                                    \
  *(mask_t*)&negativeZero = mask_t(1) << 8*sizeof(T)-1;              \
  /* Create an infinity */                                           \
  T inf(one/zero);                                                   \
  /* Create a NaN */                                                 \
  T nan(zero/zero);                                                  \
  /* Create a number of different NANs - they should    */           \
  /* all be the same on Intel chips.                    */           \
  T nan1(inf-inf);                                                   \
  T nan2(inf/inf);                                                   \
  T nan3(inf*zero);                                                  \
  T nan4(sqrt(-one));                                                \
  /* Copy one of the NANs and modify its representation. */          \
  /* This will still give a NAN, just a different one.   */          \
  T nan5(nan1);                                                      \
  (*(mask_t*)&nan5) += 1;                                            \
  /* Create number nearest to 2 (1 ULP below)            */          \
  T nearestTwo(two);                                                 \
  (*(mask_t*)&nearestTwo) -= 1;                                      \
  /* Create a number near to 2 (sizeof(T) ULPs below)     */         \
  T nearTwo(two);                                                    \
  (*(mask_t*)&nearTwo) -= sizeof(T);                                 \
  /* Create a number not so near to 2 (sizeof(T)^2 ULPs below) */    \
  T notNearTwo(two);                                                 \
  (*(mask_t*)&notNearTwo) -= sizeof(T)*sizeof(T);                    \
  /* Create a denormal by starting with zero and         */          \
  /* incrementing the integer representation.            */          \
  T smallestDenormal(0);                                             \
  (*(mask_t*)&smallestDenormal) += 1;

#define printNumber(os, x)                                           \
{ int p(2*sizeof(x)+1);                                              \
  os << setprecision(p) << left << setw(17) << #x << " = "           \
     << setw(p+6) << x << " (" << hex << showbase << setw(p+1)       \
     << *(mask_t*)&x << dec << ")" << endl;                          \
}
#define showNumbers(T)                                               \
{ LOG_INFO("showNumbers("#T")");                                     \
  printNumber(cout, zero);                                           \
  printNumber(cout, one);                                            \
  printNumber(cout, two);                                            \
  printNumber(cout, negativeZero);                                   \
  printNumber(cout, nan);                                            \
  printNumber(cout, nan1);                                           \
  printNumber(cout, nan2);                                           \
  printNumber(cout, nan3);                                           \
  printNumber(cout, nan4);                                           \
  printNumber(cout, nan5);                                           \
  printNumber(cout, nearestTwo);                                     \
  printNumber(cout, nearTwo);                                        \
  printNumber(cout, notNearTwo);                                     \
  printNumber(cout, smallestDenormal);                               \
}

#define testNegative(T)                                              \
{ LOG_INFO("testNegative("#T")");                                    \
  ASSERT(!Numeric::isNegative(zero));                                \
  ASSERT(Numeric::isNegative(negativeZero));                         \
  ASSERT(!Numeric::isNegative(one));                                 \
  ASSERT(Numeric::isNegative(-one));                                 \
  ASSERT(!Numeric::isNegative(inf));                                 \
  ASSERT(Numeric::isNegative(-inf));                                 \
}

#define testFinite(T)                                                \
{ LOG_INFO("testFinite("#T")");                                      \
  ASSERT(Numeric::isFinite(zero));                                   \
  ASSERT(Numeric::isFinite(negativeZero));                           \
  ASSERT(Numeric::isFinite(one));                                    \
  ASSERT(Numeric::isFinite(-one));                                   \
  ASSERT(!Numeric::isFinite(inf));                                   \
  ASSERT(!Numeric::isFinite(-inf));                                  \
  ASSERT(!Numeric::isFinite(nan));                                   \
  ASSERT(!Numeric::isFinite(nan1));                                  \
  ASSERT(!Numeric::isFinite(nan2));                                  \
  ASSERT(!Numeric::isFinite(nan3));                                  \
  ASSERT(!Numeric::isFinite(nan4));                                  \
  ASSERT(!Numeric::isFinite(nan5));                                  \
}

#define testInf(T)                                                   \
{ LOG_INFO("testInf("#T")");                                         \
  ASSERT(!Numeric::isInf(zero));                                     \
  ASSERT(!Numeric::isInf(one));                                      \
  ASSERT(!Numeric::isInf(-one));                                     \
  ASSERT(Numeric::isInf(inf));                                       \
  ASSERT(Numeric::isInf(-inf));                                      \
  ASSERT(!Numeric::isInf(nan));                                      \
  ASSERT(!Numeric::isInf(nan1));                                     \
  ASSERT(!Numeric::isInf(nan2));                                     \
  ASSERT(!Numeric::isInf(nan3));                                     \
  ASSERT(!Numeric::isInf(nan4));                                     \
  ASSERT(!Numeric::isInf(nan5));                                     \
  ASSERT(one/inf == 0);                                              \
  ASSERT(one/(-inf) == 0);                                           \
  ASSERT((-one)/inf == 0);                                           \
  ASSERT((-one)/(-inf) == 0);                                        \
  ASSERT(Numeric::isInf(inf*inf));                                   \
  ASSERT(Numeric::isInf(inf*(-inf)));                                \
  ASSERT(Numeric::isInf((-inf)*inf));                                \
  ASSERT(Numeric::isInf((-inf)*(-inf)));                             \
  ASSERT(Numeric::isInf(inf+inf));                                   \
  ASSERT(Numeric::isInf((-inf)+(-inf)));                             \
}

#define testNan(T)                                                   \
{ LOG_INFO("testNan("#T")");                                         \
  ASSERT(!Numeric::isNan(zero));                                     \
  ASSERT(!Numeric::isNan(one));                                      \
  ASSERT(!Numeric::isNan(-one));                                     \
  ASSERT(!Numeric::isNan(inf));                                      \
  ASSERT(!Numeric::isNan(-inf));                                     \
  ASSERT(Numeric::isNan(zero/zero));                                 \
  ASSERT(Numeric::isNan(zero/negativeZero));                         \
  ASSERT(Numeric::isNan(negativeZero/zero));                         \
  ASSERT(Numeric::isNan(negativeZero/negativeZero));                 \
  ASSERT(Numeric::isNan(inf-inf));                                   \
  ASSERT(Numeric::isNan(inf/inf));                                   \
  ASSERT(Numeric::isNan(inf/(-inf)));                                \
  ASSERT(Numeric::isNan((-inf)/inf));                                \
  ASSERT(Numeric::isNan((-inf)/(-inf)));                             \
  ASSERT(Numeric::isNan(inf*zero));                                  \
  ASSERT(Numeric::isNan((-inf)*zero));                               \
  ASSERT(Numeric::isNan(inf*negativeZero));                          \
  ASSERT(Numeric::isNan((-inf)*negativeZero));                       \
}

#define testCompare(T)                                                     \
{ LOG_INFO("testCompare("#T")");                                           \
                                                                           \
  /* The first set of tests check things that any self-respecting       */ \
  /* comparison function should agree upon.                             */ \
                                                                           \
  /* Make sure that zero and negativeZero compare as equal.             */ \
  ASSERT(Numeric::compare(zero, negativeZero, mask_t(0)));                 \
                                                                           \
  /* Make sure that nearby numbers compare as equal.                    */ \
  ASSERT(Numeric::compare(two, nearestTwo));                               \
                                                                           \
  /* Make sure that slightly more distant numbers compare as equal.     */ \
  ASSERT(Numeric::compare(two, nearTwo));                                  \
                                                                           \
  /* Make sure the results are the same with parameters reversed.       */ \
  ASSERT(Numeric::compare(nearTwo, two));                                  \
                                                                           \
  /* Make sure that even more distant numbers don't compare as equal.   */ \
  ASSERT(!Numeric::compare(two, notNearTwo));                              \
                                                                           \
  /* The next set of tests check things where the correct answer isn't  */ \
  /* as obvious or important. Some of these tests check for cases that  */ \
  /* are rare or can easily be avoided. Some of them check for cases    */ \
  /* where the behavior of the 2sComplement function is arguably better */ \
  /* than the behavior of the fussier Final function.                   */ \
                                                                           \
  /* Test wrapping from inf to -inf; maxUlps is larger than mantissa.   */ \
  /* I.e. maxUlps = 1 << 23 for floats, and 1 << 52 for doubles;        */ \
  /* or as formula: maxUlps = 1 << 8*(sizeof(T)-1)-(2*sizeof(T)/3)+1    */ \
  mask_t maxUlps = (mask_t)1 << 8*(sizeof(T)-1)-(2*sizeof(T)/3)+1;         \
  LOG_DEBUG_STR("maxUlps = " << hex << showbase << maxUlps);               \
  ASSERT(!Numeric::compare(inf, -inf, maxUlps));                           \
                                                                           \
  /* Test whether max floating point value and infinity                 */ \
  /* (representationally adjacent) compare as unequal.                  */ \
  ASSERT(!Numeric::compare(numeric_limits<T>::max(), inf));                \
                                                                           \
  /* Test whether a NAN compares as unequal to itself.                  */ \
  ASSERT(!Numeric::compare(nan, nan));                                     \
                                                                           \
  /* Test whether a NAN compares as unequal to a different NAN.         */ \
  ASSERT(!Numeric::compare(nan2, nan3));                                   \
                                                                           \
  /* Test whether tiny numbers of opposite signs compare as equal.      */ \
  ASSERT(Numeric::compare(smallestDenormal, -smallestDenormal));           \
}

#define testAll(T)                                                   \
{ initNumbers(T);                                                    \
  showNumbers(T);                                                    \
  testNegative(T);                                                   \
  testFinite(T);                                                     \
  testInf(T);                                                        \
  testNan(T);                                                        \
  testCompare(T);                                                    \
}

void testFloat()
{
  float zero(0.0);
  float one(1.0);
  float two(2.0);
  float none(zero-one);
  float pinf(one/zero);
  float ninf(none/zero);
  float nan(pinf/pinf);
  long ione(1);
  float pdmin(*(float*)(&ione)); // smallest denormalized float
  float ndmin(-pdmin);
  
  ASSERT(!Numeric::isNegative(zero));
  ASSERT(Numeric::isFinite(zero));
  ASSERT(!Numeric::isInf(zero));
  ASSERT(!Numeric::isNan(zero));

  ASSERT(!Numeric::isNegative(one));
  ASSERT(Numeric::isFinite(one));
  ASSERT(!Numeric::isInf(one));
  ASSERT(!Numeric::isNan(one));

  ASSERT(Numeric::isNegative(none));
  ASSERT(Numeric::isFinite(none));
  ASSERT(!Numeric::isInf(none));
  ASSERT(!Numeric::isNan(none));

  ASSERT(!Numeric::isNegative(pinf));
  ASSERT(!Numeric::isFinite(pinf));
  ASSERT(Numeric::isInf(pinf));
  ASSERT(!Numeric::isNan(pinf));

  ASSERT(Numeric::isNegative(ninf));
  ASSERT(!Numeric::isFinite(ninf));
  ASSERT(Numeric::isInf(ninf));
  ASSERT(!Numeric::isNan(ninf));

  ASSERT(!Numeric::isFinite(nan));
  ASSERT(!Numeric::isInf(nan));
  ASSERT(Numeric::isNan(nan));

  ASSERT(one/pinf == 0);
  ASSERT(one/ninf == 0);
  ASSERT(none/pinf == 0);
  ASSERT(none/ninf == 0);

  ASSERT(Numeric::isInf(pinf*pinf));
  ASSERT(Numeric::isInf(pinf*ninf));
  ASSERT(Numeric::isInf(ninf*pinf));
  ASSERT(Numeric::isInf(ninf*ninf));

  ASSERT(Numeric::isInf(one/zero));
  ASSERT(Numeric::isInf(none/zero));

  ASSERT(Numeric::isInf(pinf+pinf));
  ASSERT(Numeric::isInf(ninf+ninf));

  ASSERT(Numeric::isNan(zero/zero));
  ASSERT(Numeric::isNan(zero/(-zero)));
  ASSERT(Numeric::isNan((-zero)/zero));
  ASSERT(Numeric::isNan((-zero)/(-zero)));

  ASSERT(Numeric::isNan(pinf-pinf));
  ASSERT(Numeric::isNan(ninf-ninf));

  ASSERT(Numeric::isNan(pinf/pinf));
  ASSERT(Numeric::isNan(pinf/ninf));
  ASSERT(Numeric::isNan(ninf/pinf));
  ASSERT(Numeric::isNan(ninf/ninf));

  ASSERT(Numeric::isNan(pinf*zero));
  ASSERT(Numeric::isNan(ninf*zero));

  ASSERT(!Numeric::compare(pdmin, ndmin));
  ASSERT(Numeric::compare(pdmin, ndmin, 2));

  ASSERT(!Numeric::compare(one, two));
  // distance between one and two in ULPs is 0x800000L
  ASSERT(Numeric::compare(one, two, 0x800000L)); 
  ASSERT(!Numeric::compare(one, two, 0x7FFFFFL));
}

void testDouble()
{
  double zero(0.0);
  double one(1.0);
  double none(zero-one);
  double pinf(one/zero);
  double ninf(none/zero);
  double nan(pinf/pinf);

  ASSERT(!Numeric::isNegative(zero));
  ASSERT(Numeric::isFinite(zero));
  ASSERT(!Numeric::isInf(zero));
  ASSERT(!Numeric::isNan(zero));

  ASSERT(!Numeric::isNegative(one));
  ASSERT(Numeric::isFinite(one));
  ASSERT(!Numeric::isInf(one));
  ASSERT(!Numeric::isNan(one));

  ASSERT(Numeric::isNegative(none));
  ASSERT(Numeric::isFinite(none));
  ASSERT(!Numeric::isInf(none));
  ASSERT(!Numeric::isNan(none));

  ASSERT(!Numeric::isNegative(pinf));
  ASSERT(!Numeric::isFinite(pinf));
  ASSERT(Numeric::isInf(pinf));
  ASSERT(!Numeric::isNan(pinf));

  ASSERT(Numeric::isNegative(ninf));
  ASSERT(!Numeric::isFinite(ninf));
  ASSERT(Numeric::isInf(ninf));
  ASSERT(!Numeric::isNan(ninf));

  ASSERT(!Numeric::isFinite(nan));
  ASSERT(!Numeric::isInf(nan));
  ASSERT(Numeric::isNan(nan));

  ASSERT(one/pinf == 0);
  ASSERT(one/ninf == 0);
  ASSERT(none/pinf == 0);
  ASSERT(none/ninf == 0);

  ASSERT(Numeric::isInf(pinf*pinf));
  ASSERT(Numeric::isInf(pinf*ninf));
  ASSERT(Numeric::isInf(ninf*pinf));
  ASSERT(Numeric::isInf(ninf*ninf));

  ASSERT(Numeric::isInf(one/zero));
  ASSERT(Numeric::isInf(none/zero));

  ASSERT(Numeric::isInf(pinf+pinf));
  ASSERT(Numeric::isInf(ninf+ninf));

  ASSERT(Numeric::isNan(zero/zero));
  ASSERT(Numeric::isNan(zero/(-zero)));
  ASSERT(Numeric::isNan((-zero)/zero));
  ASSERT(Numeric::isNan((-zero)/(-zero)));

  ASSERT(Numeric::isNan(pinf-pinf));
  ASSERT(Numeric::isNan(ninf-ninf));

  ASSERT(Numeric::isNan(pinf/pinf));
  ASSERT(Numeric::isNan(pinf/ninf));
  ASSERT(Numeric::isNan(ninf/pinf));
  ASSERT(Numeric::isNan(ninf/ninf));

  ASSERT(Numeric::isNan(pinf*zero));
  ASSERT(Numeric::isNan(ninf*zero));
}


int main(int /*argc*/, const char* argv[])
{
  INIT_LOGGER(argv[0]);
  LOG_INFO("Starting up ...");
  try {
    testAll(float);
    testAll(double);
  }
  catch (Exception& e) {
    LOG_ERROR_STR(e);
    return 1;
  }
  LOG_INFO("Program terminated successfully");
  return 0;
}
