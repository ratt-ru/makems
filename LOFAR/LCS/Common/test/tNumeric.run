#!/bin/sh

# Skip the test until problem with gcc-4 in opt mode is resolved.
exit 3
