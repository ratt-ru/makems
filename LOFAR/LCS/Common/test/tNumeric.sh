#!/bin/sh
./runctest.sh -stdout tNumeric 2>&1 > tNumeric.log
