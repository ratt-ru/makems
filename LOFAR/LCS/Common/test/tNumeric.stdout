zero              = 0               (0         )
one               = 1               (0x3f800000)
two               = 2               (0x40000000)
negativeZero      = -0              (0x80000000)
nan               = nan             (0xffc00000)
nan1              = nan             (0xffc00000)
nan2              = nan             (0xffc00000)
nan3              = nan             (0xffc00000)
nan4              = nan             (0xffc00000)
nan5              = nan             (0xffc00001)
nearestTwo        = 1.99999988      (0x3fffffff)
nearTwo           = 1.99999952      (0x3ffffffc)
notNearTwo        = 1.99999809      (0x3ffffff0)
smallestDenormal  = 1.40129846e-45  (0x1       )
zero              = 0                       (0                 )
one               = 1                       (0x3ff0000000000000)
two               = 2                       (0x4000000000000000)
negativeZero      = -0                      (0x8000000000000000)
nan               = nan                     (0xfff8000000000000)
nan1              = nan                     (0xfff8000000000000)
nan2              = nan                     (0xfff8000000000000)
nan3              = nan                     (0xfff8000000000000)
nan4              = nan                     (0xfff8000000000000)
nan5              = nan                     (0xfff8000000000001)
nearestTwo        = 1.9999999999999998      (0x3fffffffffffffff)
nearTwo           = 1.9999999999999982      (0x3ffffffffffffff8)
notNearTwo        = 1.9999999999999858      (0x3fffffffffffffc0)
smallestDenormal  = 4.9406564584124654e-324 (0x1               )
